File: Exceções - IF Quest/python/mago.py
from personagem import Personagem
from sem_mana_error import SemManaError

class Mago(Personagem):
    def __init__(self, nome, vida, nivel):
        super().__init__(nome, vida, nivel)
        self.mana = 50

    def get_mana(self):
        return self.mana

    def set_mana(self, mana):
        if mana < 0:
            raise ValueError(f"A mana não pode ser negativa: {mana}")
        self.mana = mana

    def ficha(self):
        return super().ficha() + f"\nMana: {self.mana}"

    def habilidade(self):
        return "rajada arcana"

    def lancar_magia(self, custo):
        if custo > self.mana:
            raise SemManaError(self.mana)
        self.mana -= custo
        return f"Magia lançada com sucesso! custo: {custo}, mana restante: {self.mana}"

    def agir(self, alvo):
        resultado = self.lancar_magia(20)
        print(resultado)
        dano = self.get_nivel() * 15
        alvo.receber_dano(dano)
        print(f"{self.get_nome()} atinge {alvo.get_nome()} com magia causando {dano} de dano!")

File: Exceções - IF Quest/java/excecoesifquest/Mago.java
package excecoesifquest;

public class Mago extends Personagem {
  
    private int mana;

    public Mago(String nome, int vida, int nivel){
    super(nome, vida, nivel);
    mana = 50;
}
public int getMana(){
    return mana;
}
public void setMana(int mana){
    if (mana < 0){
        throw new IllegalArgumentException("A mana não pode ser negativa: " + mana);
    } else {
        this.mana = mana;
    }
}
@Override
public String ficha() {
    return super.ficha() + "\nMana: " + mana;
}

@Override
public String habilidade(){
    return "rajada arcana";
}

public String lancarMagia(int custo) throws SemManaException {
    if (custo > mana){
        throw new SemManaException(mana);
    } else {
        mana -= custo;
        return "Magia lançada com sucesso! custo: " + custo + ", mana restante: " + mana;
    }
}

@Override
public void agir(Personagem alvo) throws SemManaException {
    String resultado = lancarMagia(20);
    System.out.println(resultado);
    int dano = getNivel() * 15;
    alvo.receberDano(dano);
    System.out.println(getNome() + " atinge " + alvo.getNome() + " com magia causando " + dano + " de dano!");
}
}